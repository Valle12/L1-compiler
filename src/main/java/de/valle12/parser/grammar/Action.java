package de.valle12.parser.grammar;

public enum Action implements Symbol {
  FINISH
}
