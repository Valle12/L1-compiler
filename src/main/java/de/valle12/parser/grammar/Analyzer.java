package de.valle12.parser.grammar;

import de.valle12.lexer.tokens.TokenType;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;
import java.util.stream.IntStream;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public class Analyzer {
  private final List<String> productions;

  public static void main(String[] args) throws IOException {
    List<String> productions = Files.readAllLines(Path.of("src/main/resources/LL1.txt"));
    Analyzer analyzer = new Analyzer(productions);
    Map<NonTerminal, Set<TokenType>> firstSets = analyzer.createFirstSets();
    Map<NonTerminal, Set<TokenType>> followSets = analyzer.createFollowSets(firstSets);
    System.out.println("TEST");
  }

  public Map<NonTerminal, Set<TokenType>> createFirstSets() {
    Map<NonTerminal, Set<TokenType>> firstSets = new EnumMap<>(NonTerminal.class);

    for (int i = productions.size() - 1; i >= 0; i--) {
      String[] production = productions.get(i).split("->");
      NonTerminal nonTerminal = NonTerminal.valueOf(production[0].trim());

      Set<TokenType> firstSet = new HashSet<>();
      String[] symbols = production[1].split("\\|");
      for (String symbol : symbols) {
        String[] tokens = symbol.trim().split(" ");
        if (tokens.length == 1 && tokens[0].equals("ε")) {
          firstSet.add(TokenType.EPSILON);
        } else if (Character.isLowerCase(tokens[0].charAt(0))) {
          firstSet.add(TokenType.valueOf(tokens[0].toUpperCase()));
        } else {
          firstSet.addAll(firstSets.get(NonTerminal.valueOf(tokens[0])));
        }
      }

      firstSets.put(nonTerminal, firstSet);
    }

    return firstSets;
  }

  public Map<NonTerminal, Set<TokenType>> createFollowSets(
      Map<NonTerminal, Set<TokenType>> firstSets) {
    Map<NonTerminal, Set<TokenType>> followSets = initializeEmptyFollowSets();
    for (String production : productions) {
      String[] productionParts = production.split("->");
      NonTerminal nonTerminal = NonTerminal.valueOf(productionParts[0].trim());
      List<String> rhsNonTerminalProductions = getProductionsWhereNonTerminalIsOnRhs(nonTerminal);
      extractFollows(rhsNonTerminalProductions, nonTerminal, firstSets, followSets);
    }

    return followSets;
  }

  private void extractFollows(
      List<String> rhsNonTerminalProductions,
      NonTerminal rhsNonTerminal,
      Map<NonTerminal, Set<TokenType>> firstSets,
      Map<NonTerminal, Set<TokenType>> followSets) {
    for (String production : rhsNonTerminalProductions) {
      String[] productionParts = production.split("->");
      NonTerminal lhsNonTerminal = NonTerminal.valueOf(productionParts[0].trim());
      String[] symbols = productionParts[1].split("\\|"); // Could be epsilon, could be a terminal
      Set<TokenType> followSet = new HashSet<>();
      for (String symbol : symbols) {
        if (symbol.equals("ε")) continue;
        String[] tokens = symbol.trim().split(" ");
        int index =
            IntStream.range(0, tokens.length)
                .filter(i -> tokens[i].equals(rhsNonTerminal.name()))
                .findFirst()
                .orElse(-1);
        followSet.addAll(determineFollowSet(index, tokens, firstSets, followSets, lhsNonTerminal));
      }

      followSets.get(rhsNonTerminal).addAll(followSet);
    }

    if (rhsNonTerminal == NonTerminal.PROGRAM)
      followSets.get(NonTerminal.PROGRAM).add(TokenType.EOF);
  }

  private Set<TokenType> determineFollowSet(
      int index,
      String[] tokens,
      Map<NonTerminal, Set<TokenType>> firstSets,
      Map<NonTerminal, Set<TokenType>> followSets,
      NonTerminal lhsNonTerminal) {
    if (index == -1) return Set.of();
    Set<TokenType> followSet = new HashSet<>();

    if (index + 1 >= tokens.length) { // A -> pB
      followSet.addAll(followSets.get(lhsNonTerminal));
    } else if (Character.isLowerCase(
        tokens[index + 1].charAt(0))) { // A -> pBq, where q is terminal
      followSet.add(TokenType.valueOf(tokens[index + 1].trim().toUpperCase()));
    } else {
      NonTerminal nextNonTerminal = NonTerminal.valueOf(tokens[index + 1].trim());
      Set<TokenType> firstSet = new HashSet<>(firstSets.get(nextNonTerminal));
      if (firstSet.contains(TokenType.EPSILON)) {
        firstSet.remove(TokenType.EPSILON);
        firstSet.addAll(followSets.get(lhsNonTerminal));
      }

      followSet.addAll(firstSet);
    }

    return followSet;
  }

  private Map<NonTerminal, Set<TokenType>> initializeEmptyFollowSets() {
    Map<NonTerminal, Set<TokenType>> followSets = new EnumMap<>(NonTerminal.class);
    for (NonTerminal nonTerminal : NonTerminal.values()) {
      followSets.put(nonTerminal, new HashSet<>());
    }

    return followSets;
  }

  private List<String> getProductionsWhereNonTerminalIsOnRhs(NonTerminal nonTerminal) {
    return productions.stream()
        .filter(
            s -> {
              String s1 = s.split("->")[1].trim();
              return s1.contains(nonTerminal.name() + " ") || s1.endsWith(nonTerminal.name());
            })
        .toList();
  }
}
