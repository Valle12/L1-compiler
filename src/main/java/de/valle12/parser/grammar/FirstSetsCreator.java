package de.valle12.parser.grammar;

import de.valle12.lexer.tokens.TokenType;
import java.util.*;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
@Getter
public class FirstSetsCreator {
  private final List<String> productions;
  private final Map<NonTerminal, Set<TokenType>> generalFirstSets =
      new EnumMap<>(NonTerminal.class);
  private final Map<NonTerminal, RhsProduction> fineGrainedFirstSets =
      new EnumMap<>(NonTerminal.class);

  public void createFirstSets() {
    for (int i = productions.size() - 1; i >= 0; i--) {
      String[] production = productions.get(i).split("->");
      NonTerminal lhsNonTerminal = NonTerminal.valueOf(production[0].trim());
      String[] productionRules = production[1].split("\\|");
      Map<Symbol, Set<TokenType>> rhsFirstSets = new HashMap<>();
      for (String productionRule : productionRules) {
        Set<TokenType> firstSet = new HashSet<>();
        String[] tokens = productionRule.trim().split(" ");
        String token = tokens[0].trim();
        Symbol symbol;

        if (token.equals("Îµ")) {
          symbol = TokenType.EPSILON;
          firstSet.add(TokenType.EPSILON);
        } else if (Character.isLowerCase(token.charAt(0))) {
          TokenType tokenType = TokenType.valueOf(token.toUpperCase());
          symbol = tokenType;
          firstSet.add(tokenType);
        } else {
          symbol = NonTerminal.valueOf(token.toUpperCase());
          RhsProduction rhsProduction = fineGrainedFirstSets.get(symbol);
          rhsProduction.rhsProduction().values().forEach(firstSet::addAll);
        }

        rhsFirstSets.put(symbol, firstSet);
      }

      fineGrainedFirstSets.put(lhsNonTerminal, new RhsProduction(rhsFirstSets));
    }

    for (Map.Entry<NonTerminal, RhsProduction> entry : fineGrainedFirstSets.entrySet()) {
      Set<TokenType> firstSet = new HashSet<>();
      for (Set<TokenType> set : entry.getValue().rhsProduction().values()) {
        firstSet.addAll(set);
      }

      generalFirstSets.put(entry.getKey(), firstSet);
    }
  }
}
