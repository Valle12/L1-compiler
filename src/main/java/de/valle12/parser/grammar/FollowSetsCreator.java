package de.valle12.parser.grammar;

import de.valle12.lexer.tokens.TokenType;
import java.util.*;
import java.util.stream.IntStream;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
@Getter
public class FollowSetsCreator {
  private final List<String> productions;
  private final Map<NonTerminal, Set<TokenType>> firstSets;
  private final Map<NonTerminal, Set<TokenType>> followSets = initializeEmptyFollowSets();

  public void createFollowSets() {
    for (String production : productions) {
      String[] productionParts = production.split("->");
      NonTerminal nonTerminal = NonTerminal.valueOf(productionParts[0].trim());
      List<String> rhsNonTerminalProductions = getProductionsWhereNonTerminalIsOnRhs(nonTerminal);
      extractFollows(rhsNonTerminalProductions, nonTerminal);
    }
  }

  private Map<NonTerminal, Set<TokenType>> initializeEmptyFollowSets() {
    Map<NonTerminal, Set<TokenType>> sets = new EnumMap<>(NonTerminal.class);
    for (NonTerminal nonTerminal : NonTerminal.values()) {
      sets.put(nonTerminal, new HashSet<>());
    }

    return sets;
  }

  private List<String> getProductionsWhereNonTerminalIsOnRhs(NonTerminal nonTerminal) {
    return productions.stream()
        .filter(
            s -> {
              String s1 = s.split("->")[1].trim();
              return s1.contains(nonTerminal.name() + " ") || s1.endsWith(nonTerminal.name());
            })
        .toList();
  }

  private void extractFollows(List<String> rhsNonTerminalProductions, NonTerminal rhsNonTerminal) {
    for (String production : rhsNonTerminalProductions) {
      String[] productionParts = production.split("->");
      NonTerminal lhsNonTerminal = NonTerminal.valueOf(productionParts[0].trim());
      String[] symbols = productionParts[1].split("\\|"); // Could be epsilon, could be a terminal
      Set<TokenType> followSet = new HashSet<>();
      for (String symbol : symbols) {
        if (symbol.equals("Îµ")) continue;
        String[] tokens = symbol.trim().split(" ");
        int index =
            IntStream.range(0, tokens.length)
                .filter(i -> tokens[i].equals(rhsNonTerminal.name()))
                .findFirst()
                .orElse(-1);
        followSet.addAll(determineFollowSet(index, tokens, lhsNonTerminal));
      }

      followSets.get(rhsNonTerminal).addAll(followSet);
    }

    if (rhsNonTerminal == NonTerminal.PROGRAM)
      followSets.get(NonTerminal.PROGRAM).add(TokenType.EOF);
  }

  private Set<TokenType> determineFollowSet(
      int index, String[] tokens, NonTerminal lhsNonTerminal) {
    if (index == -1) return Set.of();
    Set<TokenType> followSet = new HashSet<>();

    if (index + 1 >= tokens.length) { // A -> pB
      followSet.addAll(followSets.get(lhsNonTerminal));
    } else if (Character.isLowerCase(
        tokens[index + 1].charAt(0))) { // A -> pBq, where q is terminal
      followSet.add(TokenType.valueOf(tokens[index + 1].trim().toUpperCase()));
    } else {
      NonTerminal nextNonTerminal = NonTerminal.valueOf(tokens[index + 1].trim());
      Set<TokenType> firstSet = new HashSet<>(firstSets.get(nextNonTerminal));
      if (firstSet.contains(TokenType.EPSILON)) {
        firstSet.remove(TokenType.EPSILON);
        firstSet.addAll(followSets.get(lhsNonTerminal));
      }

      followSet.addAll(firstSet);
    }

    return followSet;
  }
}
