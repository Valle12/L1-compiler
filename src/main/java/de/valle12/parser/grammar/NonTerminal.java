package de.valle12.parser.grammar;

public enum NonTerminal implements Symbol {
  PROGRAM,
  STMTS,
  STMT,
  DECL,
  SIMP,
  LVALUE,
  ASSIGN,
  ADD,
  MUL,
  UNARY,
  ASSIGN2,
  ADD2,
  MUL2,
  DECL2,
  PRIMARY,
  INTCONST,
  ASNOP,
  ADDOP,
  MULOP
}
