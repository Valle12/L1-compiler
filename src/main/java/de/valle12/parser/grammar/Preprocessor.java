package de.valle12.parser.grammar;

import de.valle12.parser.grammar.table.ParsingTable;
import de.valle12.parser.grammar.table.ParsingTableCreator;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class Preprocessor {
  @SneakyThrows
  public static void main(String[] args) {
    LOGGER.info("Calculate first and follow sets...");
    List<String> productions = Files.readAllLines(Path.of("src/main/resources/LL1.txt"));

    FirstSetsCreator firstSetsCreator = new FirstSetsCreator(productions);
    firstSetsCreator.createFirstSets();
    LOGGER.info("First sets created successfully.");

    FollowSetsCreator followSetsCreator =
        new FollowSetsCreator(productions, firstSetsCreator.getGeneralFirstSets());
    followSetsCreator.createFollowSets();
    LOGGER.info("Follow sets created successfully.");

    ParsingTableCreator parsingTableCreator =
        new ParsingTableCreator(
            productions,
            firstSetsCreator.getFineGrainedFirstSets(),
            followSetsCreator.getFollowSets());
    parsingTableCreator.createParsingTable();
    LOGGER.info("Parsing table created successfully.");

    ParsingTable parsingTable = parsingTableCreator.getParsingTable();
    parsingTable.saveTableToFile();
    parsingTable.exportParsingTableToCsv();
    LOGGER.info("Parsing table saved to file and exported to CSV successfully.");
  }
}
