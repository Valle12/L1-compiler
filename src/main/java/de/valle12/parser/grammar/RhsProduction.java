package de.valle12.parser.grammar;

import de.valle12.lexer.tokens.TokenType;
import java.util.Map;
import java.util.Set;

public record RhsProduction(Map<Symbol, Set<TokenType>> rhsProduction) {}
