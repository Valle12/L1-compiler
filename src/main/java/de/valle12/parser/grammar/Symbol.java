package de.valle12.parser.grammar;

public interface Symbol {}
