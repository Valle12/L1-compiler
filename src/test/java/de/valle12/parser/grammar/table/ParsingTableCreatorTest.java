package de.valle12.parser.grammar.table;

import de.valle12.parser.grammar.FirstSetsCreator;
import de.valle12.parser.grammar.FollowSetsCreator;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import lombok.SneakyThrows;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

class ParsingTableCreatorTest {
  @Test
  @DisplayName("Test createParsingTable with valid input")
  @SneakyThrows
  void test1() {
    List<String> productions = Files.readAllLines(Path.of("src/main/resources/LL1.txt"));
    FirstSetsCreator firstSetsCreator = new FirstSetsCreator(productions);
    firstSetsCreator.createFirstSets();
    FollowSetsCreator followSetsCreator =
        new FollowSetsCreator(productions, firstSetsCreator.getGeneralFirstSets());
    followSetsCreator.createFollowSets();
    ParsingTableCreator parsingTableCreator =
        new ParsingTableCreator(
            productions,
            firstSetsCreator.getFineGrainedFirstSets(),
            followSetsCreator.getFollowSets());

    parsingTableCreator.createParsingTable();

    Approvals.verify(parsingTableCreator.getParsingTable());
  }
}
