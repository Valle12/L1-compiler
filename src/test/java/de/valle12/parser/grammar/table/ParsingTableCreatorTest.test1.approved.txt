ParsingTable[table={PROGRAM=ParsingTableRow[row={CLASS=[ParsingTableProduction[symbols=[CLASS, MAIN, LEFT_PARENTHESIS, RIGHT_PARENTHESIS, LEFT_BRACE, STMTS, RIGHT_BRACE]]]}], STMTS=ParsingTableRow[row={RETURN=[ParsingTableProduction[symbols=[STMT, STMTS]]], CLASS=[ParsingTableProduction[symbols=[STMT, STMTS]]], LEFT_PARENTHESIS=[ParsingTableProduction[symbols=[STMT, STMTS]]], IDENTIFIER=[ParsingTableProduction[symbols=[STMT, STMTS]]], RIGHT_BRACE=[ParsingTableProduction[symbols=[EPSILON]]], EPSILON=[ParsingTableProduction[symbols=[EPSILON]]]}], STMT=ParsingTableRow[row={RETURN=[ParsingTableProduction[symbols=[RETURN, ASSIGN, SEMICOLON]]], CLASS=[ParsingTableProduction[symbols=[DECL, SEMICOLON]]], LEFT_PARENTHESIS=[ParsingTableProduction[symbols=[SIMP, SEMICOLON]]], IDENTIFIER=[ParsingTableProduction[symbols=[SIMP, SEMICOLON]]]}], DECL=ParsingTableRow[row={CLASS=[ParsingTableProduction[symbols=[CLASS, IDENTIFIER, DECL2]]]}], SIMP=ParsingTableRow[row={LEFT_PARENTHESIS=[ParsingTableProduction[symbols=[LVALUE, ASNOP, ASSIGN]]], IDENTIFIER=[ParsingTableProduction[symbols=[LVALUE, ASNOP, ASSIGN]]]}], LVALUE=ParsingTableRow[row={LEFT_PARENTHESIS=[ParsingTableProduction[symbols=[LEFT_PARENTHESIS, LVALUE, RIGHT_PARENTHESIS]]], IDENTIFIER=[ParsingTableProduction[symbols=[IDENTIFIER]]]}], ASSIGN=ParsingTableRow[row={LEFT_PARENTHESIS=[ParsingTableProduction[symbols=[ADD, ASSIGN2]]], MINUS=[ParsingTableProduction[symbols=[ADD, ASSIGN2]]], IDENTIFIER=[ParsingTableProduction[symbols=[ADD, ASSIGN2]]], DECIMAL=[ParsingTableProduction[symbols=[ADD, ASSIGN2]]], HEXADECIMAL=[ParsingTableProduction[symbols=[ADD, ASSIGN2]]]}], ADD=ParsingTableRow[row={LEFT_PARENTHESIS=[ParsingTableProduction[symbols=[MUL, ADD2]]], MINUS=[ParsingTableProduction[symbols=[MUL, ADD2]]], IDENTIFIER=[ParsingTableProduction[symbols=[MUL, ADD2]]], DECIMAL=[ParsingTableProduction[symbols=[MUL, ADD2]]], HEXADECIMAL=[ParsingTableProduction[symbols=[MUL, ADD2]]]}], MUL=ParsingTableRow[row={LEFT_PARENTHESIS=[ParsingTableProduction[symbols=[UNARY, MUL2]]], MINUS=[ParsingTableProduction[symbols=[UNARY, MUL2]]], IDENTIFIER=[ParsingTableProduction[symbols=[UNARY, MUL2]]], DECIMAL=[ParsingTableProduction[symbols=[UNARY, MUL2]]], HEXADECIMAL=[ParsingTableProduction[symbols=[UNARY, MUL2]]]}], UNARY=ParsingTableRow[row={LEFT_PARENTHESIS=[ParsingTableProduction[symbols=[PRIMARY]]], MINUS=[ParsingTableProduction[symbols=[MINUS, UNARY]]], IDENTIFIER=[ParsingTableProduction[symbols=[PRIMARY]]], DECIMAL=[ParsingTableProduction[symbols=[PRIMARY]]], HEXADECIMAL=[ParsingTableProduction[symbols=[PRIMARY]]]}], ASSIGN2=ParsingTableRow[row={SEMICOLON=[ParsingTableProduction[symbols=[EPSILON]]], EQUALS=[ParsingTableProduction[symbols=[ASNOP, ASSIGN]]], RIGHT_PARENTHESIS=[ParsingTableProduction[symbols=[EPSILON]]], PLUS_EQUALS=[ParsingTableProduction[symbols=[ASNOP, ASSIGN]]], MINUS_EQUALS=[ParsingTableProduction[symbols=[ASNOP, ASSIGN]]], STAR_EQUALS=[ParsingTableProduction[symbols=[ASNOP, ASSIGN]]], SLASH_EQUALS=[ParsingTableProduction[symbols=[ASNOP, ASSIGN]]], PERCENT_EQUALS=[ParsingTableProduction[symbols=[ASNOP, ASSIGN]]], EPSILON=[ParsingTableProduction[symbols=[EPSILON]]]}], ADD2=ParsingTableRow[row={SEMICOLON=[ParsingTableProduction[symbols=[EPSILON]]], EQUALS=[ParsingTableProduction[symbols=[EPSILON]]], RIGHT_PARENTHESIS=[ParsingTableProduction[symbols=[EPSILON]]], PLUS=[ParsingTableProduction[symbols=[ADDOP, MUL, ADD2]]], MINUS=[ParsingTableProduction[symbols=[ADDOP, MUL, ADD2]]], PLUS_EQUALS=[ParsingTableProduction[symbols=[EPSILON]]], MINUS_EQUALS=[ParsingTableProduction[symbols=[EPSILON]]], STAR_EQUALS=[ParsingTableProduction[symbols=[EPSILON]]], SLASH_EQUALS=[ParsingTableProduction[symbols=[EPSILON]]], PERCENT_EQUALS=[ParsingTableProduction[symbols=[EPSILON]]], EPSILON=[ParsingTableProduction[symbols=[EPSILON]]]}], MUL2=ParsingTableRow[row={SEMICOLON=[ParsingTableProduction[symbols=[EPSILON]]], EQUALS=[ParsingTableProduction[symbols=[EPSILON]]], RIGHT_PARENTHESIS=[ParsingTableProduction[symbols=[EPSILON]]], PLUS=[ParsingTableProduction[symbols=[EPSILON]]], MINUS=[ParsingTableProduction[symbols=[EPSILON]]], STAR=[ParsingTableProduction[symbols=[MULOP, UNARY, MUL2]]], SLASH=[ParsingTableProduction[symbols=[MULOP, UNARY, MUL2]]], PERCENT=[ParsingTableProduction[symbols=[MULOP, UNARY, MUL2]]], PLUS_EQUALS=[ParsingTableProduction[symbols=[EPSILON]]], MINUS_EQUALS=[ParsingTableProduction[symbols=[EPSILON]]], STAR_EQUALS=[ParsingTableProduction[symbols=[EPSILON]]], SLASH_EQUALS=[ParsingTableProduction[symbols=[EPSILON]]], PERCENT_EQUALS=[ParsingTableProduction[symbols=[EPSILON]]], EPSILON=[ParsingTableProduction[symbols=[EPSILON]]]}], DECL2=ParsingTableRow[row={SEMICOLON=[ParsingTableProduction[symbols=[EPSILON]]], EQUALS=[ParsingTableProduction[symbols=[EQUALS, ASSIGN]]], EPSILON=[ParsingTableProduction[symbols=[EPSILON]]]}], PRIMARY=ParsingTableRow[row={LEFT_PARENTHESIS=[ParsingTableProduction[symbols=[LEFT_PARENTHESIS, ASSIGN, RIGHT_PARENTHESIS]]], IDENTIFIER=[ParsingTableProduction[symbols=[IDENTIFIER]]], DECIMAL=[ParsingTableProduction[symbols=[INTCONST]]], HEXADECIMAL=[ParsingTableProduction[symbols=[INTCONST]]]}], INTCONST=ParsingTableRow[row={DECIMAL=[ParsingTableProduction[symbols=[DECIMAL]]], HEXADECIMAL=[ParsingTableProduction[symbols=[HEXADECIMAL]]]}], ASNOP=ParsingTableRow[row={EQUALS=[ParsingTableProduction[symbols=[EQUALS]]], PLUS_EQUALS=[ParsingTableProduction[symbols=[PLUS_EQUALS]]], MINUS_EQUALS=[ParsingTableProduction[symbols=[MINUS_EQUALS]]], STAR_EQUALS=[ParsingTableProduction[symbols=[STAR_EQUALS]]], SLASH_EQUALS=[ParsingTableProduction[symbols=[SLASH_EQUALS]]], PERCENT_EQUALS=[ParsingTableProduction[symbols=[PERCENT_EQUALS]]]}], ADDOP=ParsingTableRow[row={PLUS=[ParsingTableProduction[symbols=[PLUS]]], MINUS=[ParsingTableProduction[symbols=[MINUS]]]}], MULOP=ParsingTableRow[row={STAR=[ParsingTableProduction[symbols=[STAR]]], SLASH=[ParsingTableProduction[symbols=[SLASH]]], PERCENT=[ParsingTableProduction[symbols=[PERCENT]]]}]}]